symbolic-base/src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Optimization.hs

```diff
@@ -0,0 +1,93 @@
+module ZkFold.Symbolic.Compiler.ArithmeticCircuit.Optimization (
+        optimize,
+        isLinear,
+        toConstVar
+    ) where
+
+import           Data.Bool                                               (bool)
+import           Data.Functor.Rep                                        (Representable (..))
+import           Data.Map                                                hiding (drop, foldl, foldr, map, null, splitAt,
+                                                                          take)
+import qualified Data.Map.Internal                                       as M
+import qualified Data.Map.Monoidal                                       as MM
+import qualified Data.Set                                                as S
+import           Numeric.Natural                                         (Natural)
+import           Prelude                                                 hiding (Num (..), drop, length, product,
+                                                                          splitAt, sum, take, (!!), (^))
+
+import           ZkFold.Base.Algebra.Basic.Class
+import           ZkFold.Base.Algebra.Polynomials.Multivariate.Monomial   (Mono (..))
+import           ZkFold.Base.Algebra.Polynomials.Multivariate.Polynomial (Poly (..), polynomial)
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Instance     ()
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Internal
+import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Witness      (WitnessF (..))
+
+--------------------------------- High-level functions --------------------------------
+
+-- | Replaces linear polynoms of the form
+-- @(fromConstant k) * (NewVar nV) + (fromConstant c)@
+-- with a constant variable @fromConstant $ negate c // k@ in an arithmetic circuit
+-- and replaces variable with a constant in witness
+--
+optimize ::
+  ( Arithmetic a, Ord (Rep i)) =>
+  ArithmeticCircuit a p i o  -> ArithmeticCircuit a p i o
+optimize ac = case [toConstVar p | (_, p) <- toList (acSystem ac), isLinear p] of
```

Comment:
Current performance is ok I guess, but if we need it to be faster later, we can do the following:
1. Notice that the optimization passes depend only on the state of `acSystem`
2. Instead of updating the whole circuit on each pass, we can only update the constraint system, accumulating the list of constant variables
3. After the system stabilizes (no more constants can be inferred), we can apply the whole accumulated list of constant variables to the rest of the circuit instead of applying it one-by-one